### PR TITLE
The four extensions in one package "extras"

### LINKED ISSUE #12: Add extras extension package for insert, mark, subscript, and superscript
I would like to add these inline extensions as additional options to Hugo’s Goldmark instances.

We have been discussing the first two ones in the [Forum](https://discourse.gohugo.io/t/extensions-for-super-and-subscripts/48150). 

These extensions follow the same syntax as the ones for markdown-it:
- `1^st^` &rightarrow; `1<sup>st</sup>` (see [npm repo](https://www.npmjs.com/package/markdown-it-sup))
- `H~2~O` &rightarrow; `H<sub>2</sub>O` (see [npm repo](https://www.npmjs.com/package/markdown-it-sub)
- `++insert++` &rightarrow; `<ins>insert</ins>` (see [npm repo](https://www.npmjs.com/package/markdown-it-ins)

Their default configuration switch should be set to false, because every new extension adds more load to the parser and the demand is not very high. 

My fork of this repo already contains the code for these extensions, including many test cases. I tried to follow the structure of the first extension `passthrough` and took your advice in #8 into consideration.

I already build a local Hugo version including these extensions and would know how to integrate them.

### PR DESCRIPTION
**Pull request by bowman2001**
Closes #12

### CODE REVIEW

**Comment by jmooring** [comment]
The example from the README contains this:

```text
Hydrogen (H) is the 1^st^ element in the periodic table.

Water (H~2~O) is a liquid.

Water (H~2~O) is a ++transparent++ liquid.
```

The subscripts are failing.

This works:

```text
x H~2~O
```

This does not:

```text
x H~2~O x
```

File: extras/subscript.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/yuin/goldmark/renderer/html"
+	"github.com/yuin/goldmark/text"
+	"github.com/yuin/goldmark/util"
+)
+
+type subscriptDelimiterProcessor struct {
+}
+
+func (p *subscriptDelimiterProcessor) IsDelimiter(b byte) bool {
+	return b == '~'
```
Comment (bep) [inline extras/subscript.go:18]: This is a general comment that spans all insert/subscript/superscript.go files:

I have skimmed the implementation of all of these 3 and they look almost exactly the same, the only diff seems to be the delimiter and the HTML wrapper element.

I don't mind code duplication, but in this case I think we can make everything simpler by wrapping all of these into one extension/parser/renderer.

File: extras/subscript.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+var defaultSubscriptDelimiterProcessor = &subscriptDelimiterProcessor{}
+
+type subscriptParser struct {
+}
+
+var defaultSubscriptParser = &subscriptParser{}
+
+// NewSubscriptParser returns a new InlineParser that parses subscript expressions.
+func NewSubscriptParser() parser.InlineParser {
```
Comment (bep) [inline extras/subscript.go:37]: This is related to the above (one parser), but I suggest that we for starters unexports everything that's not related to getting a extension.

File: extras/subscript.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		_, _ = w.WriteString("</sub>")
+	}
+	return gast.WalkContinue, nil
+}
+
+type subscript struct {
+}
+
+// Subscript is an extension that allows you to use a subscript expression like 'H~2~O'.
+var Subscript = &subscript{}
```
Comment (bep) [inline extras/subscript.go:106]: This works, but there will (maybe) eventually come some config, and I suspect that this becomes easier to document across packages if we have a conistent API. So I suggest we do something ala:

```go
extensions := extras.New(
    extras.SubscriptConfig{
            Enable: true
     },
     extras.SuperscriptConfig{
            Enable: true
     },
     ....
}
```

With my comment above about unexporting things, we could probably add something ala this later:

```go
extensions := extras.New(
    extras.SubscriptConfig{
            Enable: true
            HTMLRenderer: ...
     },
     ....
}
```

**Review by bep — CHANGES_REQUESTED** [review]

**Comment by bowman2001** [comment]
Yes, @bep, the only difference between Superscript and Subscript are the markers (a single '^' or '~'). All the other variables, functions, or methods only have different names. This redundancy could surly be removed. 

The insert doesn’t need to watch out for whitespace and has the same logic as strong emphasis. 

But first I need to address the problem @jmooring mentioned. My unit tests did not cover the cases "x H~2~O x" or "x foo^3^bar x" until now. They both fail, and I obviously missed something I need to figure out.

File: extras/go.mod [hunk 4]
```diff
@@ -0,0 +1,5 @@
+module github.com/gohugoio/hugo-goldmark-extensions/extras
+
+go 1.22.1
```
Comment (jmooring) [inline extras/go.mod:3]:
```suggestion
go 1.20
```

**Comment by bowman2001** [comment]
Sorry for the delay, I had to attend to a lot of changes in my life in the last weeks. Coming month my new full-time Job is going to start. 

I want to finish the extras package with the requested changes in the next days.

File: extras/subscript.go (hunk 3 above)
Comment (bowman2001) [reply extras/subscript.go]: I'm using an enumeration type to identify and configure the extra inline tags with internal structs. I hope this is in the line of what you meant.

**Comment by bowman2001** [comment]
There are new unit tests for the problem @jmooring found, and it's gone. The redundancy in the extensions is also gone. Only the necessary general structures for an inline tag remain which can be used for different tags. I have added a new mark tag because it was so easy: ==marked== -> <mark>marked</mark>. 

The configuration follows now the same concept as `passthrough` without any additional configuration data, as @bep requested.

I also changed `config.go` and `convert.go` in my Hugo fork and wrote some integration tests in `goldmark_integration_test.go`. Everything looks good.

I noticed that I am not as proficient with the Go workflow as I would like. I had to change the module path to my personal repo to integrate.

**Comment by jmooring** [comment]
To simplify integration testing I am merging this without tagging a release. The repository is still described as "Work in progress".

**Pull request merged**